In [104]:
# import library
import pandas as pd
import numpy as np

In [2]:
# load the datasets inti a dataframe
df=pd.read_csv('google_play_apps.csv')

In [3]:
# first 5 rows
df.head()

Unnamed: 0                                                App  \
0           0     Photo Editor & Candy Camera & Grid & ScrapBook   
1           1                                Coloring book moana   
2           2  U Launcher Lite – FREE Live Cool Themes, Hide ...   
3           3                              Sketch - Draw & Paint   
4           4              Pixel Draw - Number Art Coloring Book   

         Category  Rating  Reviews  Size     Installs  Type Price  \
0  ART_AND_DESIGN     4.1      159  19.0      10,000+  Free     0   
1  ART_AND_DESIGN     3.9      967  14.0     500,000+  Free     0   
2  ART_AND_DESIGN     4.7    87510   8.7   5,000,000+  Free     0   
3  ART_AND_DESIGN     4.5   215644  25.0  50,000,000+  Free     0   
4  ART_AND_DESIGN     4.3      967   2.8     100,000+  Free     0   

  Content Rating                     Genres      Last Updated  \
0       Everyone               Art & Design   January 7, 2018   
1       Everyone  Art & Design;Pretend Play  January 15, 2018   
2       Everyone               Art & Design    August 1, 2018   
3           Teen               Art & Design      June 8, 2018   
4       Everyone    Art & Design;Creativity     June 20, 2018   

          Current Ver   Android Ver  
0               1.0.0  4.0.3 and up  
1               2.0.0  4.0.3 and up  
2               1.2.4  4.0.3 and up  
3  Varies with device    4.2 and up  
4                 1.1    4.4 and up

In [4]:
# drop one column from the datasets because we don't need that column in future
df.drop('Unnamed: 0',axis=1,inplace=True)

In [5]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159  19.0      10,000+  Free     0       Everyone   
1      967  14.0     500,000+  Free     0       Everyone   
2    87510   8.7   5,000,000+  Free     0       Everyone   
3   215644  25.0  50,000,000+  Free     0           Teen   
4      967   2.8     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [7]:
df.shape

(9659, 13)

In [8]:
# all information about datasets
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9659 non-null   object 
 1   Category        9659 non-null   object 
 2   Rating          8196 non-null   float64
 3   Reviews         9659 non-null   int64  
 4   Size            8432 non-null   float64
 5   Installs        9659 non-null   object 
 6   Type            9659 non-null   object 
 7   Price           9659 non-null   object 
 8   Content Rating  9659 non-null   object 
 9   Genres          9659 non-null   object 
 10  Last Updated    9659 non-null   object 
 11  Current Ver     9651 non-null   object 
 12  Android Ver     9657 non-null   object 
dtypes: float64(2), int64(1), object(10)
memory usage: 981.1+ KB


## Data Cleaning
The four features that we will be working with most frequently henceforth are Installs, Size, Rating and Price. The info() function (from the previous task) told us that Installs and Price columns are of type object and not int64 or float64 as we would expect. This is because the column contains some characters more than just [0,9] digits. Ideally, we would want these columns to be numeric as their name suggests.
Hence, we now proceed to data cleaning and prepare our data to be consumed in our analyis later. Specifically, the presence of special characters (, $ +) in the Installs and Price columns make their conversion to a numerical data type difficult

In [9]:
df['Installs']=df['Installs'].astype(str).str.replace('[^0-9]','')

<ipython-input-9-c843937a01d2>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Installs']=df['Installs'].astype(str).str.replace('[^0-9]','')


In [10]:
df['Installs']=pd.to_numeric(df['Installs'])

In [12]:
df['Price']=df['Price'].astype(str).str.replace('[^0-9]','')

<ipython-input-12-45df9a316d53>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Price']=df['Price'].astype(str).str.replace('[^0-9]','')


In [13]:
df['Price']=pd.to_numeric(df['Price'])

In [14]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159  19.0     10000  Free      0       Everyone   
1      967  14.0    500000  Free      0       Everyone   
2    87510   8.7   5000000  Free      0       Everyone   
3   215644  25.0  50000000  Free      0           Teen   
4      967   2.8    100000  Free      0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [15]:
# check null values in datasets
df.isnull().sum()

App                  0
Category             0
Rating            1463
Reviews              0
Size              1227
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

In [16]:
df=df[df['Size'].notnull()]

In [17]:
df=df[df['Current Ver'].notnull()]

In [18]:
df=df[df['Android Ver'].notnull()]

In [19]:
df.isnull().sum()

App                  0
Category             0
Rating            1401
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
dtype: int64

1. How many unique categories are present in the dataset?

In [22]:
len(df['Category'].unique())

33

2. Fill the null values in Rating column depending on the Category column

In [23]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159  19.0     10000  Free      0       Everyone   
1      967  14.0    500000  Free      0       Everyone   
2    87510   8.7   5000000  Free      0       Everyone   
3   215644  25.0  50000000  Free      0           Teen   
4      967   2.8    100000  Free      0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [24]:
grpc=df.groupby('Category')

In [27]:
for i in df['Category'].unique():
    df['Rating'][df['Category']==i]=df['Rating'][df['Category']==i].fillna(grpc['Rating'].mean()[i])

<ipython-input-27-441d7ea55064>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Rating'][df['Category']==i]=df['Rating'][df['Category']==i].fillna(grpc['Rating'].mean()[i])


In [28]:
df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

In [29]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159  19.0     10000  Free      0       Everyone   
1      967  14.0    500000  Free      0       Everyone   
2    87510   8.7   5000000  Free      0       Everyone   
3   215644  25.0  50000000  Free      0           Teen   
4      967   2.8    100000  Free      0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

3. Create a binned column using the Rating column which should contain 3 bins namely less than 3, 3 to 4, 4 to 5

In [31]:
df['Rating_bin']=pd.cut(df['Rating'],bins=(0,3,4,5),labels=('0 t0 less than 3','3 to 4','4 to 5'))

In [32]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159  19.0     10000  Free      0       Everyone   
1      967  14.0    500000  Free      0       Everyone   
2    87510   8.7   5000000  Free      0       Everyone   
3   215644  25.0  50000000  Free      0           Teen   
4      967   2.8    100000  Free      0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver Rating_bin  
0  4.0.3 and up     4 to 5  
1  4.0.3 and up     3 to 4  
2  4.0.3 and up     4 to 5  
3    4.2 and up     4 to 5  
4    4.4 and up     4 to 5

4. Name the app which was updated recently

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8422 entries, 0 to 9658
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   App             8422 non-null   object  
 1   Category        8422 non-null   object  
 2   Rating          8422 non-null   float64 
 3   Reviews         8422 non-null   int64   
 4   Size            8422 non-null   float64 
 5   Installs        8422 non-null   int64   
 6   Type            8422 non-null   object  
 7   Price           8422 non-null   int64   
 8   Content Rating  8422 non-null   object  
 9   Genres          8422 non-null   object  
 10  Last Updated    8422 non-null   object  
 11  Current Ver     8422 non-null   object  
 12  Android Ver     8422 non-null   object  
 13  Rating_bin      8422 non-null   category
dtypes: category(1), float64(2), int64(3), object(8)
memory usage: 1.2+ MB


In [35]:
# covert the data type of 'Last Updated' column
df['Last Updated']=pd.to_datetime(df['Last Updated'])

In [37]:
# sort the value by 'Last Updated' column in descending order
df.sort_values(by='Last Updated',ascending=False,inplace=True)

In [42]:
# recently updated Apps
df[df['Last Updated']=='2018-08-08'].reset_index(drop=True)

App            Category  Rating  \
0                                    BankNordik             FINANCE     3.9   
1                       Shoot Hunter-Gun Killer                GAME     4.3   
2                               Fast Tract Diet  HEALTH_AND_FITNESS     4.4   
3  Video Downloader For FB: Save FB Videos 2018               TOOLS     4.3   

   Reviews  Size  Installs  Type  Price Content Rating            Genres  \
0       28  15.0      5000  Free      0       Everyone           Finance   
1   320334  27.0  50000000  Free      0           Teen            Action   
2       35   2.4      1000  Paid    799       Everyone  Health & Fitness   
3      434   3.6     50000  Free      0       Everyone             Tools   

  Last Updated Current Ver   Android Ver Rating_bin  
0   2018-08-08       7.3.2    5.0 and up     3 to 4  
1   2018-08-08       1.1.2    4.1 and up     4 to 5  
2   2018-08-08       1.9.3    4.2 and up     4 to 5  
3   2018-08-08       1.1.9  4.0.3 and up     4 to 5

5. How many apps based on content rating which Size are less than 10.

In [50]:
no_of_app=df[df['Size']<10].groupby('Content Rating')['App'].count()
print('no of app based on content rating are less than 10: ',no_of_app)

no of app based on content rating are less than 10:  Content Rating
Adults only 18+       1
Everyone           3361
Everyone 10+         45
Mature 17+          118
Teen                258
Unrated               2
Name: App, dtype: int64


In [51]:
total_no_app=no_of_app.sum()
print('Total no of app based on content rating are less than 10: ',total_no_app)

Total no of app based on content rating are less than 10:  3785


6. How many teen apps belong to action genres

In [102]:
df.reset_index(drop=True,inplace=True)

In [103]:
df.head()

App            Category  Rating  \
0                                    BankNordik             FINANCE     3.9   
1                       Shoot Hunter-Gun Killer                GAME     4.3   
2                               Fast Tract Diet  HEALTH_AND_FITNESS     4.4   
3  Video Downloader For FB: Save FB Videos 2018               TOOLS     4.3   
4                  Baby Food - Homemade Recipes           PARENTING     4.1   

   Reviews  Size  Installs  Type  Price Content Rating            Genres  \
0       28  15.0      5000  Free      0       Everyone           Finance   
1   320334  27.0  50000000  Free      0           Teen            Action   
2       35   2.4      1000  Paid    799       Everyone  Health & Fitness   
3      434   3.6     50000  Free      0       Everyone             Tools   
4       44   3.6     10000  Free      0       Everyone         Parenting   

  Last Updated Current Ver   Android Ver Rating_bin  
0   2018-08-08       7.3.2    5.0 and up     3 to 4  
1   2018-08-08       1.1.2    4.1 and up     4 to 5  
2   2018-08-08       1.9.3    4.2 and up     4 to 5  
3   2018-08-08       1.1.9  4.0.3 and up     4 to 5  
4   2018-08-07         8.2  4.0.3 and up     4 to 5

In [94]:
# from below operation we found the no of 'teen' App belong to 'Genres'='Action'
df[df['Genres']=='Action'].groupby('Content Rating').count()

App  Category  Rating  Reviews  Size  Installs  Type  Price  \
Content Rating                                                                
Everyone         62        62      62       62    62        62    62     62   
Everyone 10+     46        46      46       46    46        46    46     46   
Mature 17+       39        39      39       39    39        39    39     39   
Teen            134       134     134      134   134       134   134    134   

                Genres  Last Updated  Current Ver  Android Ver  Rating_bin  
Content Rating                                                              
Everyone            62            62           62           62          62  
Everyone 10+        46            46           46           46          46  
Mature 17+          39            39           39           39          39  
Teen               134           134          134          134         134

Answer  of the Above ques. is : total 134 Teen App belongs to 'Action Genres'

In [101]:
df[df['Content Rating']=='Teen'].groupby('Genres').count()

App  Category  Rating  Reviews  Size  Installs  Type  \
Genres                                                                          
Action                   134       134     134      134   134       134   134   
Adventure                 29        29      29       29    29        29    29   
Arcade                    23        23      23       23    23        23    23   
Art & Design               3         3       3        3     3         3     3   
Auto & Vehicles            1         1       1        1     1         1     1   
Beauty                     3         3       3        3     3         3     3   
Board                      3         3       3        3     3         3     3   
Books & Reference          8         8       8        8     8         8     8   
Business                  13        13      13       13    13        13    13   
Card                      21        21      21       21    21        21    21   
Casino                    31        31      31       31    31        31    31   
Casual                    23        23      23       23    23        23    23   
Comics                    16        16      16       16    16        16    16   
Communication             23        23      23       23    23        23    23   
Dating                     9         9       9        9     9         9     9   
Education                 14        14      14       14    14        14    14   
Educational               13        13      13       13    13        13    13   
Entertainment            102       102     102      102   102       102   102   
Events                     6         6       6        6     6         6     6   
Finance                    5         5       5        5     5         5     5   
Food & Drink               4         4       4        4     4         4     4   
Health & Fitness          13        13      13       13    13        13    13   
House & Home               2         2       2        2     2         2     2   
Lifestyle                 22        22      22       22    22        22    22   
Maps & Navigation          2         2       2        2     2         2     2   
Medical                    5         5       5        5     5         5     5   
Music                      2         2       2        2     2         2     2   
News & Magazines          24        24      24       24    24        24    24   
Parenting                  1         1       1        1     1         1     1   
Personalization           49        49      49       49    49        49    49   
Photography               10        10      10       10    10        10    10   
Productivity               7         7       7        7     7         7     7   
Puzzle                    17        17      17       17    17        17    17   
Racing                     6         6       6        6     6         6     6   
Role Playing              43        43      43       43    43        43    43   
Shopping                  19        19      19       19    19        19    19   
Simulation                31        31      31       31    31        31    31   
Social                    86        86      86       86    86        86    86   
Sports                    13        13      13       13    13        13    13   
Strategy                  33        33      33       33    33        33    33   
Tools                      5         5       5        5     5         5     5   
Travel & Local             3         3       3        3     3         3     3   
Trivia                     4         4       4        4     4         4     4   
Video Players & Editors   14        14      14       14    14        14    14   
Weather                    2         2       2        2     2         2     2   
Word                       1         1       1        1     1         1     1   

                         Price  Content Rating  Last Updated  Current Ver  \
Genres                                                     